Rewrite color scheme logic

LeaVerou opened the pull request:
_(Extracted from another PR to keep it small)_

- Minimize needless swaps
- Centralize logic
- Use an actual `<wa-select>` rather than hacking it with `<wa-dropdown>`

LeaVerou committed:
Rewrite color scheme logic

- Minimize needless swaps
- Centralize logic
- Use an actual `<wa-select>` rather than hacking it with `<wa-dropdown>`

LeaVerou committed:
Rewrite color scheme logic

- Minimize needless swaps
- Centralize logic
- Use an actual `<wa-select>` rather than hacking it with `<wa-dropdown>`

claviska commented:
This is great! The only thing I see off is this:

![CleanShot 2025-01-09 at 17 14 58@2x](https://github.com/user-attachments/assets/c17ef58a-af47-4f8e-b751-ee88b581dd45)

It doesn't happen to all the labels. I can only get it to break this one for some reason.

LeaVerou commented:
I can just not set the width, but then they are a bit too long. Any idea why this happens? It looks like a bug with `<wa-select>`.

lindsaym-fa commented:
A couple of things I've noticed in testing —

I imagine this was intentionally left out of this logic, but for awareness, the theme name and description are missing from the Themes page:

https://github.com/user-attachments/assets/8564a233-d9c1-430f-af37-561539cbfe45


(I understand this page may be retired, but this is broken behavior while the page continues to exist.)


I can't consistently get into this state to give steps to reproduce, but occasionally, the entirety of the selected theme isn't applied (colors and component-specific overrides seem to still work). Here's a video in Firefox, but this also occurred for me in Safari:

https://github.com/user-attachments/assets/fb94af91-5a93-4e3e-9f25-e2cd031faa51

LeaVerou commented:
> I imagine this was intentionally left out of this logic, but for awareness, the theme name and description are missing from the Themes page:

Nice catch! Damn, breaking a big change into smaller commits is _really_ hard. Not sure how to best approach this, #477 depends on this, and now this depends on it :( 

> I can't consistently get into this state to give steps to reproduce, but occasionally, the entirety of the selected theme isn't applied (colors and component-specific overrides seem to still work). Here's a video in Firefox, but this also occurred for me in Safari:

Interesting. I wonder if it's a bug with Turbo. Presumably it's gone when you refresh, right?

lindsaym-fa commented:
> Presumably it's gone when you refresh, right?

A hard refresh usually solves it. A normal refresh does not.

It looks like maybe the last locally stored preset theme is lingering in the `<head>`? Maybe?
<img width="751" alt="Screenshot 2025-01-09 at 5 19 39 PM" src="https://github.com/user-attachments/assets/6f8cc5c6-057c-46c3-923d-c98129520598" />

ETA: In this attempt, I selected 'Awesome', refreshed the page, then selected 'Brutalist'.
<img width="692" alt="Screenshot 2025-01-09 at 5 23 15 PM" src="https://github.com/user-attachments/assets/85eaeb9f-185a-4219-84b6-616e60e602bc" />

LeaVerou committed:
Merge branch 'theme-pickers' of https://github.com/shoelace-style/webawesome into theme-pickers

LeaVerou commented:
Interesting. What is logged when you run `localStorage.presetTheme` in the console? Does this happen with the color scheme as well or only the theme?

lindsaym-fa commented:
> What is logged when you run localStorage.presetTheme in the console? Does this happen with the color scheme as well or only the theme?

I would love to tell you, but I can't get back into the right state because our options are missing. 😅
<img width="481" alt="Screenshot 2025-01-09 at 5 28 13 PM" src="https://github.com/user-attachments/assets/5be91a4c-98f8-4758-b607-49b23c2ddd12" />

LeaVerou commented:
Whoa, what happened there? How did this happen?

LeaVerou commented:
@claviska This is how it looks with no explicit widths, no idea why:
<img width="508" alt="image" src="https://github.com/user-attachments/assets/dfdcd305-7ddf-4e98-9366-14b288201a75" />

LeaVerou committed:
Update theme across windows

lindsaym-fa commented:
> Whoa, what happened there? How did this happen?

Not exactly sure... I'm missing icons, the theme selector, and color scheme selector on the Vercel preview in every browser. Tried to refresh, hard refresh, clear cache... No good. 🤔

LeaVerou committed:
Remove unused `*-wide` templates

LeaVerou committed:
ALL THE THINGS

- Move shared includes to shared template
- Rewrite critical theme JS (from 67 down to 20 loc)
- Refactor of theme picker code
- Utilize VT better and use them for color scheme too
- Do not store default value in localStorage

LeaVerou committed:
Remove widths

LeaVerou committed:
Fix theme showcase regression

LeaVerou committed:
Fix #482

LeaVerou commented:
Alright, I just pushed a bunch of changes!

- Move shared includes to shared template
- Rewrite critical theme JS (from 67 down to 20 loc)
- Refactor of theme picker code
- Utilize view transitions better and use them for color scheme too
- Do not store default value in localStorage
- Removed unused `*-wide` templates
- Removed explicit widths (hopefully one of you could help debug these select widths)
- Fixed theme showcase regression PLUS existing bug #482 

Could you please take another look?

claviska commented:
look I'm all about minimalism but this is ridiculous 😆 

![CleanShot 2025-01-10 at 09 27 40@2x](https://github.com/user-attachments/assets/5e726c19-2076-4eea-baa4-78b9966e0388)

LeaVerou commented:
Can you reproduce locally? I think it only happens in the Vercel preview

LeaVerou commented:
@claviska it looks like the SSR loader is failing to load in the Vercel preview, which seems unrelated to this PR?

LeaVerou committed:
Merge branch 'next' into theme-pickers
